src/collapse/bpftrace.rs

```diff
@@ -0,0 +1,213 @@
+                            }
+                            // Write the current character to last string in our "stack"
+                            vec.last_mut().unwrap().push(c);
+                        }
+                    }
+                }
+                if is_beginning_of_line {
+                    is_beginning_of_line = false;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+enum State {
+    NotInStack,
+    InStack(Vec<String>),
+}
+
+impl Default for State {
+    fn default() -> Self {
+        State::NotInStack
+    }
+}
+
+/// Consumes all whitespace, if any, until the beginning of a series of digits. Then consumes
+/// the series of digits, writing their ascii representation into the provided buffer. Returns
+/// the number of bytes written.
+fn consume_unsigned_integer<I>(buf: &mut [u8], chars: &mut Peekable<I>) -> Result<usize, io::Error>
+where
+    I: Iterator<Item = char>,
+{
+    // We're expecting a number; so if we run out of characters, something went wrong.
+    let mut next = match chars.peek() {
+        Some(next) => next,
+        None => {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Parsing Error. Expected a number. Found '\n'.",
+            ));
+        }
+    };
+
+    // Consume and ignore all whitespace. Again, we're expecting a number; so if we run out
+    // of characters, something went wrong.
+    while next.is_whitespace() {
+        chars.next().unwrap();
+        next = match chars.peek() {
+            Some(c) => c,
+            None => {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Parsing Error. Expected a number. Found '\n'.",
+                ));
+            }
+        }
+    }
+
+    // Consume all digits, writing their ascii representation into the provided buffer
+    let mut index = 0;
+    while next.is_numeric() {
+        let c = chars.next().unwrap();
+        buf[index] = c as u8;
+        index += 1;
+        next = match chars.peek() {
+            Some(c) => c,
+            None => &'a',
+        };
+    }
+
+    // Return the number of bytes written
+    Ok(index)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::env;
+    use std::fs;
+    use std::io::{self, Read};
+    use std::path::Path;
+
+    use super::*;
+
+    #[test]
+    fn test_bpftrace() {
+        let root_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+        let data_dir = Path::new(&root_dir)
+            .join("tests")
+            .join("data")
+            .join("bpftrace");
+
+        for entry in fs::read_dir(&data_dir).unwrap() {
+            let path = entry.unwrap().path();
+            if let Some(extension) = path.extension() {
+                if extension.to_str().unwrap() == "trace" {
+                    let inpath = path;
+                    let outpath = inpath.with_extension("orig.folded");
```

Comment:
I may be missing something here, but I thought the intention was that the `orig.folded` files were generated by the upstream `stackcollapse-bpftrace`? Looking at them more closely, this doesn't seem to be the case? I think we specifically want to compare against the "true" results from the Perl version to make sure we conform correctly :)

Reply:
The `orig.folded` files are generated by the perl implementation (`stackcollapse-bpftrace.pl`). This is also relevant for your `trim()` comment above.

Reply:
Wow... That's super weird.. I wonder how that works. Maybe they haven't noticed because the SVG collapses whitespace in strings? Judging from the example input they give at the top of `stackcollapse-bpftrace`, it looks like they just assume that there's no indentation in the stack lines.. Let me file an issue there and see what they say!

Reply:
Even though the `orig.folded` files in the data directory *are* generated by the upstream perl implementation, we could if you'd like have the test run the perl script each time it runs. Let me know if you would prefer that. I originally wrote the tests that way, but then changed for no apparent reason. So just let me know which one you prefer.

Reply:
Oooh, I like that. That'd be even better! It'd slow down the tests a little, but I think it's okay.

Reply:
Also, I just filed https://github.com/brendangregg/FlameGraph/issues/201

Reply:
👍